pkg/proxy/backup.go

```diff
@@ -0,0 +1,387 @@
+		os.Setenv(part[0], part[1])
+	}
+
+	recv, err := backupstore.List(req.VolumeName, req.DestUrl, req.VolumeOnly)
+	if err != nil {
+		return resp, err
+	}
+
+	for k, v := range recv {
+		resp.Volumes[k] = &rpc.EngineBackupVolumeInfo{
+			Name:                 v.Name,
+			Size:                 v.Size,
+			Labels:               v.Labels,
+			Created:              v.Created,
+			LastBackupName:       v.LastBackupName,
+			LastBackupAt:         v.LastBackupAt,
+			DataStored:           v.DataStored,
+			Messages:             parseMessages(v.Messages),
+			Backups:              parseBackups(v.Backups),
+			BackingImageName:     v.BackingImageName,
+			BackingImageChecksum: v.BackingImageChecksum,
+		}
+	}
+
+	ackResponse(resp.ProxyEngineResponse, true)
+	return resp, nil
+}
+
+func parseBackups(in map[string]*backupstore.BackupInfo) (out map[string]*rpc.EngineBackupInfo) {
+	out = map[string]*rpc.EngineBackupInfo{}
+	for k, v := range in {
+		out[k] = parseBackup(v)
+	}
+	return out
+}
+
+func parseBackup(in *backupstore.BackupInfo) (out *rpc.EngineBackupInfo) {
+	return &rpc.EngineBackupInfo{
+		Name:                   in.Name,
+		Url:                    in.URL,
+		SnapshotName:           in.SnapshotName,
+		SnapshotCreated:        in.SnapshotCreated,
+		Created:                in.Created,
+		Size:                   in.Size,
+		Labels:                 in.Labels,
+		IsIncremental:          in.IsIncremental,
+		VolumeName:             in.VolumeName,
+		VolumeSize:             in.VolumeSize,
+		VolumeCreated:          in.VolumeCreated,
+		VolumeBackingImageName: in.VolumeBackingImageName,
+		Messages:               parseMessages(in.Messages),
+	}
+}
+
+func parseMessages(in map[backupstore.MessageType]string) (out map[string]string) {
+	out = map[string]string{}
+	for k, v := range in {
+		out[string(k)] = v
+	}
+	return out
+}
+
+func (p *Proxy) BackupConfigMetaGet(ctx context.Context, req *rpc.EngineBackupConfigMetaGetRequest) (resp *rpc.EngineBackupConfigMetaGetProxyResponse, err error) {
+	log := logrus.WithFields(logrus.Fields{"destURL": req.DestUrl})
+	log.Debug("Getting backup config metadata")
+
+	resp = &rpc.EngineBackupConfigMetaGetProxyResponse{
+		ProxyEngineResponse: &rpc.ProxyEngineResponse{},
+	}
+	ackResponse(resp.ProxyEngineResponse, false)
+
+	// TODO
+
+	ackResponse(resp.ProxyEngineResponse, true)
+	return resp, nil
+}
+
+func (p *Proxy) BackupRemove(ctx context.Context, req *rpc.EngineBackupRemoveRequest) (resp *rpc.ProxyEngineResponse, err error) {
+	log := logrus.WithFields(logrus.Fields{"destURL": req.DestUrl})
+	if req.VolumeName != "" {
+		log = log.WithField("volume", req.VolumeName)
+	}
+	log.Debug("Removing backups")
+
+	resp = &rpc.ProxyEngineResponse{}
+	ackResponse(resp, false)
+
+	for _, env := range req.Envs {
+		part := strings.SplitN(env, "=", 2)
+		os.Setenv(part[0], part[1])
+	}
+
+	if req.VolumeName == "" {
+		if err := backupstore.DeleteDeltaBlockBackup(req.DestUrl); err != nil {
```

Comment:
Need to use `destURL = util.UnescapeURL(destURL)` the same as backupstore command implementation.

Reply:
Not quite sure why. This is not inputted from bash.
https://github.com/longhorn/longhorn-engine/blob/master/pkg/util/util.go#L248

Reply:
yea, the point is:

As per our discussion, to avoid any missing logic when copying the code from engine/backup commands to proxy functions, I would suggest moving the shared part from commands to shared packages which can then be leveraged by commands and proxy, so it will definitely ensure nothing missing in proxy implementation. 

This is quite important, because all volume/snapshot/backup operations will be through the proxy later, so we should ensure the result is the same as running commands. 

Please create a ticket to improve it in the next sprint. Thanks.

Reply:
https://github.com/longhorn/longhorn/issues/3981